### PR TITLE
optimize BE memory gc with smooth step function

### PR DESCRIPTION
**Pull request by starrocks-xupeng**
for #183 
use jemalloc gc method to help BE do memory gc.

### CODE REVIEW

File: be/src/common/daemon.cpp [hunk 1]
```diff
@@ -98,10 +106,22 @@ void* tcmalloc_gc_thread(void* dummy) {
...
             size_t max_free_size = phy_size * config::tc_free_memory_rate / 100;
             if (free_size > max_free_size) {
-                MallocExtension::instance()->ReleaseToSystem(free_size - max_free_size);
+                total_bytes_to_gc = free_size - max_free_size;
+            }
+        }
+        size_t bytes_to_gc = gch.bytes_should_gc(MonoTime::Now(), total_bytes_to_gc);
+        if (bytes_to_gc > 0) {
+            size_t bytes = bytes_to_gc;
+            while (bytes >= GCBYTES_ONE_STEP) {
```
Comment (dirtysalt) [inline be/src/common/daemon.cpp:119]: What if this value is larger? How about 64MB or 128MB？Is there bad impact?

**Review by dirtysalt — DISMISSED** [review]

File: be/src/util/gc_helper.cpp [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+    _bytes_limit = _backlog_bytes_limit();
+    _remained_bytes = (_bytes_limit > current_bytes) ? _bytes_limit : current_bytes;
+
+    return current_bytes > _bytes_limit ? (current_bytes - _bytes_limit) : 0;
+}
+
+// move `_backlog` elements `nadvance` step forward,
+// and set `_backlog[last]` based on `current_bytes`
+void GCHelper::_backlog_update(const uint64_t nadvance, const size_t current_bytes) {
+    if (nadvance >= SMOOTHSTEP_NSTEPS) {
```
Comment (kangkaisen) [inline be/src/util/gc_helper.cpp:173]: I think this if else branch need comment

File: be/src/util/gc_helper.cpp [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+#include "util/gc_helper.h"
+
+namespace starrocks {
+
+#define SMOOTHSTEP_VARIANT "smoother"
+#define SMOOTHSTEP_BFP 24
+// generated by redis smoothstep.sh
+#define SMOOTHSTEP                                                    \
+    /* STEP(step, h,                            x,     y) */          \
+    STEP(1, UINT64_C(0x000000000000005f), 0.008, 0.000005714940202)   \
```
[the remark this answers is not in the document]
Comment (starrocks-xueli) [reply be/src/util/gc_helper.cpp:12]: they are constant already, actually we only use column 2 `h` to calculate, if we need to keep each interval even, we still need to generate an array. we just use this auto-generated to make sure there is no correctness issue.

[the remark this answers is not in the document]
Comment (starrocks-xueli) [reply be/src/util/gc_helper.cpp:12]: you can check this out, https://en.wikipedia.org/wiki/Smoothstep.

File: be/src/common/daemon.cpp (hunk 1 above)
Comment (starrocks-xueli) [reply be/src/common/daemon.cpp:119]: just an value i think is not too small or too big, setting it bigger will hold the lock longer.

**Review by trueeyu — DISMISSED** [review]

**Review by trueeyu — APPROVED** [review]

**Review by kangkaisen — APPROVED** [review]

File: be/src/util/gc_helper.cpp [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+        return 0;
+
+    delta = MonoDelta::FromNanoseconds(_interval * nadvance);
+    _epoch += delta;
+
+    // step 2: update backlog accroding to steps `nadvance`
+    _backlog_update(nadvance, current_bytes);
+
+    // step 3: compute bytes limit according to backlog
+    _bytes_limit = _backlog_bytes_limit();
```
Comment (dirtysalt) [inline be/src/util/gc_helper.cpp:37]: this variable could be local.

**Pull request merged**